qutebrowser/browser/webengine/webenginesettings.py

```diff
@@ -162,6 +182,30 @@ class WebEngineSettings(websettings.AbstractSettings):
         QWebEngineSettings.FantasyFont: QFont.Fantasy,
     }
 
+    def set_unknown_url_scheme_policy(self, policy: str) -> bool:
+        """Set the UnknownUrlSchemePolicy to use.
+
+        Return:
+            True if there was a change, False otherwise.
+        """
+        if policy is usertypes.UNSET:
+            return False
```

Comment:
I'm pretty sure this handling is incorrect - when navigating from a page with a customized URL scheme to a page without such a customization, nothing will be set now.

Reply:
I had some false assumptions before. After digging into the code, I realized this problem happened because after QWebEngineView is created, `WebEngineSettings._settings` will be changed from `_SettingsWrapper` to `QWebEngineSettings`. So `return False` when `policy is usertypes.UNNET` will not update the QWebEngineSettings,

```python
# qutebrowser/qutebrowser/browser/webengine/webenginetab.py
class WebEngineTab(browsertab.AbstractTab):
    def __init__(self, *, win_id, mode_manager, private, parent=None):
        ...
        widget = webview.WebEngineView(tabdata=self.data, win_id=win_id,
                                       private=private)
        ...
        # We're assigning settings in _set_widget
        self.settings = webenginesettings.WebEngineSettings(settings=None)
        self._set_widget(widget)

    def _set_widget(self, widget: QWidget) -> None:
        # pylint: disable=protected-access
        self._widget = widget
        ...
        self.settings._settings = widget.settings()
```

Reply:
> After digging into the code, I realized this problem happened because after QWebEngineView is created, `WebEngineSettings._settings` will be changed from `_SettingsWrapper` to `QWebEngineSettings`.

I don't think so.

For settings specific to a tab (rather than global settings), `WebEngineSettings._settings` first is `None` (as set in the constructor); then as soon as the `QWebEngineView` is known, it's set to the `QWebEngineSettings` object.

For global settings, it's always the `_SettingsWrapper()`, in the global `webenginesettings.global_settings` instance of `WebEngineSettings`. The reason it exists is because there can be multiple profiles (default/private) which both have their own setting objects, but we want to apply all settings on both of them.

Then, depending on what changes, the code flow is different:

- For globally changed settings, `webenginesettings._update_settings` (the function) calls `global_settings.update_setting(...)` (which then calls `AbstractSettings._update_setting(...)` with the value, which then uses the `_SettingsWrapper`).
- For per-domain settings, the respective tab calls `AbstractSettings.update_for_url(...)`, which then calls `AbstractSettings._update_setting(...)`, which uses the underlying setting object directly (rather than the wrapper, because there is only one setting object it needs to change).

Reply:
Thank you for the detailed explaination! The reason why this problem happend is quite clear to me now.